tests/test_env.py

```diff
@@ -1,20 +1,105 @@
+        ),
+        pytest.param(
+            {},
+            "[pytest]\nenv = MAGIC=alpha\n SORCERY=beta",
+            {"MAGIC": "alpha", "SORCERY": "beta"},
+            id="two new keys - add to env",
+        ),
+        pytest.param(
+            # This test also tests for non-interference of env variables between this test and tests above
+            {},
+            "[pytest]\nenv = d:MAGIC=beta",
+            {"MAGIC": "beta"},
+            id="D flag - add to env",
+        ),
+        pytest.param(
+            {"MAGIC": "alpha"},
+            "[pytest]\nenv = MAGIC=beta",
+            {"MAGIC": "beta"},
+            id="key exists in env - overwrite",
+        ),
+        pytest.param(
+            {"MAGIC": "alpha"},
+            "[pytest]\nenv = D:MAGIC=beta",
+            {"MAGIC": "alpha"},
+            id="D exists - original val kept",
+        ),
+        pytest.param(
+            {"PLANET": "world"},
+            "[pytest]\nenv = MAGIC=hello_{PLANET}",
+            {"MAGIC": "hello_world"},
+            id="curly exist - interpolate var",
+        ),
+        pytest.param(
+            {"PLANET": "world"},
+            "[pytest]\nenv = R:MAGIC=hello_{PLANET}",
+            {"MAGIC": "hello_{PLANET}"},
+            id="R exists - not interpolate var",
+        ),
+        pytest.param(
+            {"MAGIC": "a"},
+            "[pytest]\nenv = R:MAGIC={MAGIC}b\n D:MAGIC={MAGIC}c\n MAGIC={MAGIC}d",
+            {"MAGIC": "{MAGIC}bd"},
+            id="incremental interpolation",
+        ),
+        pytest.param(
+            {},
+            "[pytest]\nenv = __PYTEST1__:__PYTEST2__:D:RESULT=result",
+            {"RESULT": "result_success_both_flags_found"},
+            id="two flags",
+        ),
+        pytest.param(
+            {},
+            "[pytest]\nenv = __PYTEST2__:__PYTEST1__:RESULT=reversed",
+            {"RESULT": "reversed_success_both_flags_found"},
+            id="two flags - reversed",
+        ),
+        pytest.param(
+            {},
+            "[pytest]\nenv = __pytest1__:__pytest2__:RESULT=lower",
+            {"RESULT": "lower_success_both_flags_found"},
+            id="lowercase flags",
+        ),
+        pytest.param(
+            {},
+            "[pytest]\nenv =  __pytest1__  :  __pytest2__  :  RESULT  =  spaces",
+            {"RESULT": "spaces_success_both_flags_found"},
+            id="whitespace is ignored",
+        ),
+        pytest.param(
+            {"MAGIC": "zero"},
+            "",
+            {"MAGIC": "zero"},
+            id="empty ini works",
+        ),
+    ],
+)
+def test_env(testdir: pytest.Testdir, env: dict[str, str], ini: str, expected_env: dict[str, str],
+             request: pytest.FixtureRequest) -> None:
+    env = {**env, "_TEST_ENV": repr(expected_env)}
+    # monkeypatch persists env variables across parametrized tests, therefore using an alternative approach:
+    with mock.patch.dict(os.environ, env, clear=True):
```

Comment:
monkeypatch persists env variables from different tests. Things that had the same issue:
- `pytest.MonkeyPatch().setenv`
- `pytest.MonkeyPatch.setenv`
- `dirtest.monkeypatch.setenv`
- `monkeypatch` fixture
- `with monkeypatch.context()`

probably it's because monkeypatch works at the function scope, while the scope for parametrization is even smaller than that, so the function scope is never left between running different parameters.